ddtrace/internal/telemetry/metrics.py

```diff
@@ -0,0 +1,178 @@
+from ddtrace.internal.telemetry.constants import TELEMETRY_METRIC_TYPE_DISTRIBUTIONS
+from ddtrace.internal.telemetry.constants import TELEMETRY_METRIC_TYPE_GAUGE
+from ddtrace.internal.telemetry.constants import TELEMETRY_METRIC_TYPE_RATE
+
+
+MetricType = Text
+MetricTagType = Dict[str, Any]
+
+
+class Metric(six.with_metaclass(abc.ABCMeta)):
+    """
+    Telemetry Metrics are stored in DD dashboards, check the metrics in datadoghq.com/metric/explorer
+    """
+
+    metric_type = ""
+    _points = []  # type: List[Tuple[float, float]]
+
+    def __init__(self, namespace, name, tags, common, interval=None):
+        # type: (str, str, MetricTagType, bool, Optional[float]) -> None
+        """
+        namespace: the scope of the metric: tracer, appsec, etc.
+        name: string
+        tags: extra information attached to a metric
+        common: set to True if a metric is common to all tracers, false if it is python specific
+        interval: field set for gauge and rate metrics, any field set is ignored for count metrics (in secs)
+        """
+        self.name = name
+        self.is_common_to_all_tracers = common
+        self.interval = interval
+        self.namespace = namespace
+        self._tags = tags  # type: MetricTagType
+        self._count = 0.0
+        self._points = []  # type: List[float]
+
+    @property
+    def id(self):
+        """
+        https://www.datadoghq.com/blog/the-power-of-tagged-metrics/#whats-a-metric-tag
+        """
+        return self.name + str(self._tags)
+
+    def __hash__(self):
+        return self.id
+
+    @abc.abstractmethod
+    def add_point(self, value=1.0):
+        # type: (float) -> None
+        """adds timestamped data point associated with a metric"""
+        pass
+
+    def set_tags(self, tags):
+        # type: (MetricTagType) -> None
+        """sets a metrics tag"""
+        if tags:
+            for k, v in iter(tags.items()):
+                self.set_tag(k, v)
+
+    def set_tag(self, name, value):
+        # type: (str, str) -> None
+        """sets a metrics tag"""
+        self._tags[name] = value
+
+    def to_dict(self):
+        # type: () -> Dict
+        """returns a dictionary containing the metrics fields expected by the telemetry intake service"""
+        data = {
+            "metric": self.name,
+            "type": self.metric_type,
+            "common": self.is_common_to_all_tracers,
+            "points": self._points,
+            "tags": ["%s:%s" % (k, v) for k, v in self._tags.items()],
+        }
+        if self.interval is not None:
+            data["interval"] = int(self.interval)
+        return data
+
+
+class CountMetric(Metric):
+    """
+    A count type adds up all the submitted values in a time interval. This would be suitable for a
+    metric tracking the number of website hits, for instance.
+    """
+
+    metric_type = TELEMETRY_METRIC_TYPE_COUNT
+    _points = []
```

Comment:
_points is an attribute in the parent class (Metric). We shouldn't override it here.

```suggestion
```